### PR TITLE
feat: remove TODOS from users loadable "loading" state

### PR DESCRIPTION
**Pull request by thiagodallacqua-hpe**
## Description

Ticket: [WEB-905](https://determinedai.atlassian.net/browse/WEB-905)

We have quite a few TODOS in our codebase, so, this PR intends to resolve a few. The idea is to submit batches of PR's to help merging them faster.

The topic for this batch is around the usage of `Loadable.match`, especially with `useUsers`.

## Test Plan

The overall usage shouldn't have changed, only a few fallbacks with `<Spinner />` were added, still, here are the pages that should be checked (a quick refresh or just arriving at the page should do).

- Cluster
- JobQueue
- ModelDetails
- ModelRegistry
- ExperimentList
- ModelVersionDetails
- TrialsComparison and table
- WorkspaceDetails
- WorkspaceProjects
- WorkspaceList
- Projects (with the cards view enabled)
- Tasks

## Checklist

- [ ] Changes have been manually QA'd
- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
  See [Release Note](https://github.com/determined-ai/determined/blob/master/docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

## Ticket

[WEB-905]: https://determinedai.atlassian.net/browse/WEB-905?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

**Review by keita-determined — COMMENTED** [review]
i dont think we need to re-wrap data with Loadable. That requires to lots of code changes, but i dont think it is necessary.

i think its much simpler to do 

```tsx
const [isLoading, setIsLoading] = useSet<boolean>(false):
const _data = Loadable.match(useData(), {
    Loaded: (cUser) => {
        setIsLoading(false);
        return  Loaded(cUser.users)
    },
    NotLoaded: () => {
      setIsLoading(true);
      return data's default value
    },
});
```

**Comment by ClaireNeveu** [comment]
Keita, your suggestion is bad react code, you're synchronously setting state there which is something you should not do.

File: webui/react/src/components/TaskList.tsx [hunk 1]
```diff
@@ -654,7 +654,7 @@ const TaskList: React.FC<Props> = ({ workspace }: Props) => {
           ContextMenu={TaskActionDropdownCM}
           dataSource={filteredTasks}
           defaultColumns={stgsConfig.settings.columns.defaultValue}
-          loading={tasks === undefined || !settings}
+          loading={tasks === undefined || !settings || !columns.length}
```
[the remark this answers is not in the document]
Comment (ClaireNeveu) [reply webui/react/src/components/TaskList.tsx:657]: To suggest a remedy here, you can use `Loadable.map` when defining `columns`.

File: webui/react/src/components/TaskList.tsx [hunk 2]
```diff
@@ -99,18 +99,15 @@ const filterKeys: Array<keyof Settings> = ['search', 'state', 'type', 'user', 'w
 
 const TaskList: React.FC<Props> = ({ workspace }: Props) => {
   const users = Loadable.match(useUsers(), {
-    Loaded: (cUser) => cUser.users,
-    NotLoaded: () => [],
-  }); // TODO: handle loading state
+    Loaded: (cUser) => Loaded(cUser.users),
```
[the remark this answers is not in the document]
Comment (ClaireNeveu) [reply webui/react/src/components/TaskList.tsx:102]: This does nothing to keep TS happy.

[the remark this answers is not in the document]
Comment (ClaireNeveu) [reply webui/react/src/components/TaskList.tsx]: The usage of `match` here was confusing, I see now what you were trying to do and you can use `Loadable.map` for that.

**Comment by trentwatt** [comment]
I thought the TODO in these cases was to somehow immediately handle the loading state instead of propagating nullish values, but I might have misunderstood. @ClaireNeveu, can you clarify?

**Comment by ClaireNeveu** [comment]
@trentwatt I'm not sure I understand your question. Loading state should be handled in the component that loads the data and should be distinguishable from the empty state.

**Comment by trentwatt** [comment]
Ok, I thought you had meant to get rid of things like `NotLoaded: () => undefined`, but maybe I had misunderstood

**Comment by ClaireNeveu** [comment]
We should generally get rid of things like `NotLoaded: () => undefined` unless the loading state is for the thing to not appear in the page.

File: webui/react/src/components/UserSelectFilter.tsx [hunk 3]
```diff
@@ -57,8 +54,8 @@ const UserSelectFilter: React.FC<Props> = ({ onChange, value }: Props) => {
       );
     }
 
-    if (users) {
-      const allOtherUsers = users
+    if (Loadable.isLoaded(users)) {
```
Comment (ClaireNeveu) [inline webui/react/src/components/UserSelectFilter.tsx:57]: You're not actually handling the loading state here, we're still just showing an empty Select.

**Review by trentwatt — COMMENTED** [review]
This one is a bit tricky. We want to move more towards explicit and granular handling of loading states everywhere, but right now it is a common pattern in our codebase to have a mixture of boolean/aggregated loading states, i.e. 

```
const isLoading = thingALoading || thingBLoading || ...
```

whereas we want to have more like

```
return (
    <div> 
        <Spinner loading={Loadable.isLoaded(thingA)}>
             <ThingARenderer thingA={thingA} />
        </Spinner>
        <Spinner loading={Loadable.isLoaded(thingB)}>
             <ThingBRenderer thingB={thingB} />
        </Spinner>
    </div>
  )
```


When it comes to "handling loading state" for these particular loading states in line with the new approach, it is difficult to do so without changing a lot of the existing code it is intertwined with. 


That said, I think it's a good start and I had some suggestions. 

I also noticed a page crash when refreshing the page on netlify, see video. I called out the place in the code below that I think is relevant, although it doesnt appear to be the root cause.

https://user-images.githubusercontent.com/28383080/223683286-6b5149f1-4a3b-47a9-942c-2da6652b13d0.mov

File: webui/react/src/components/UserSelect.tsx [hunk 4]
```diff
@@ -2,7 +2,7 @@ import React, { useCallback, useMemo } from 'react';
```
Comment (trentwatt) [inline webui/react/src/components/UserSelect.tsx:2]: it looks like this file is not used

File: webui/react/src/pages/Cluster/ClusterHistoricalUsage.tsx [hunk 5]
```diff
@@ -136,7 +136,7 @@ const ClusterHistoricalUsage: React.FC = () => {
             />
           )}
         </Section>
-        <Section bodyBorder loading={!chartSeries} title="Compute Hours by User">
+        <Section bodyBorder loading={Loadable.isLoading(users)} title="Compute Hours by User">
```
Comment (trentwatt) [inline webui/react/src/pages/Cluster/ClusterHistoricalUsage.tsx:139]: maybe leave `!chartSeries` as well?

File: webui/react/src/components/TaskList.tsx [hunk 6]
```diff
@@ -463,14 +466,15 @@ const TaskList: React.FC<Props> = ({ workspace }: Props) => {
         dataIndex: 'user',
         defaultWidth: DEFAULT_COLUMN_WIDTHS['user'],
         filterDropdown: userFilterDropdown,
```
Comment (trentwatt) [inline webui/react/src/components/TaskList.tsx:468]: ideally we could make it where we provide a loading state to the dropdown, but i dont think it's worth it when we're in the process of switching tables

File: webui/react/src/pages/TrialsComparison/Table/TrialTable.tsx [hunk 7]
```diff
@@ -489,47 +490,47 @@ const TrialTable: React.FC<Props> = ({
...
-        startTimeColumn,
-        endTimeColumn,
-        ...hpColumns,
-        actionColumn,
-      ].map((col) => ({ ...col, dataIndex: col.key } as ColumnDef<V1AugmentedTrial>)),
-    [
-      actionColumn,
-      idColumn,
+  const columns = useMemo(() => {
+    if (userColumn === undefined) return [];
```
Comment (trentwatt) [inline webui/react/src/pages/TrialsComparison/Table/TrialTable.tsx:494]: pretty sure `userColumn` is always defined now, so we can get rid of this

File: webui/react/src/pages/WorkspaceDetails/WorkspaceProjects.tsx [hunk 8]
```diff
@@ -176,6 +175,8 @@ const WorkspaceProjects: React.FC<Props> = ({ workspace, id, pageRef }) => {
   }, []);
 
   const columns = useMemo(() => {
+    if (Loadable.isLoading(users)) return [];
```
Comment (trentwatt) [inline webui/react/src/pages/WorkspaceDetails/WorkspaceProjects.tsx:178]: this is doing thing where it returns empty columns, which i think is what was causing the page crash on the experiment list before

**Review by trentwatt — APPROVED** [review]
Mostly good, few comments

**Pull request merged**